README.rst
HTTPie: human-friendly CLI HTTP client for the API era
######################################################

HTTPie (pronounced *aitch-tee-tee-pie*) is a command-line HTTP client.
Its goal is to make CLI interaction with web services as human-friendly as possible.
HTTPie is designed for testing, debugging, and generally interacting with APIs & HTTP servers.
The ``http`` & ``https`` commands allow for creating and sending arbitrary HTTP requests.
They use simple and natural syntax and provide formatted and colorized output.



.. class:: no-web no-pdf

    |docs| |pypi| |build| |coverage| |downloads| |gitter|


.. class:: no-web no-pdf

    .. image:: https://raw.githubusercontent.com/httpie/httpie/master/httpie.gif
        :alt: HTTPie in action
        :width: 100%
        :align: center


.. contents::

.. section-numbering::


About this document
===================

This documentation is best viewed at `httpie.org/docs <https://httpie.org/docs>`_.

You can select your corresponding HTTPie version as well as run examples directly from the
browser using a `termible.io <https://termible.io?utm_source=httpie-readme>`_ embedded terminal.

If you are reading this on GitHub, then this text covers the current *development* version.
You are invited to submit fixes and improvements to the the docs by editing
`README.rst <https://github.com/httpie/httpie/blob/master/README.rst>`_.


Main features
=============

* Expressive and intuitive syntax
* Formatted and colorized terminal output
* Built-in JSON support
* Forms and file uploads
* HTTPS, proxies, and authentication
* Arbitrary request data
* Custom headers
* Persistent sessions
* Wget-like downloads
* Linux, macOS and Windows support
* Plugins
* Documentation
* Test coverage


.. class:: no-web

    .. image:: https://raw.githubusercontent.com/httpie/httpie/master/httpie.png
        :alt: HTTPie compared to cURL
        :width: 100%
        :align: center


Installation
============


macOS
-----


On macOS, HTTPie can be installed via `Homebrew <https://brew.sh/>`_
(recommended):

.. code-block:: bash

    $ brew install httpie


A MacPorts *port* is also available:

.. code-block:: bash

    $ port install httpie

Linux
-----

Most Linux distributions provide a package that can be installed using the
system package manager, for example:

.. code-block:: bash

    # Debian, Ubuntu, etc.
    $ apt install httpie

.. code-block:: bash

    # Fedora
    $ dnf install httpie

.. code-block:: bash

    # CentOS, RHEL, ...
    $ yum install httpie

.. code-block:: bash

    # Gentoo
    $ emerge httpie

.. code-block:: bash

    # Arch Linux
    $ pacman -S httpie

.. code-block:: bash

    # Solus
    $ eopkg install httpie


Windows, etc.
-------------

A universal installation method (that works on Windows, Mac OS X, Linux, …,
and always provides the latest version) is to use `pip`_:


.. code-block:: bash

    # Make sure we have an up-to-date version of pip and setuptools:
    $ python -m pip install --upgrade pip setuptools

    $ python -m pip install --upgrade httpie


(If ``pip`` installation fails for some reason, you can try
``easy_install httpie`` as a fallback.)


Python version
--------------

Python version 3.6 or greater is required.


Unstable version
----------------

You can also install the latest unreleased development version directly from
the ``master`` branch on GitHub.  It is a work-in-progress of a future stable
release so the experience might be not as smooth.


.. class:: no-pdf

|build|


On macOS you can install it with Homebrew:

.. code-block:: bash

    $ brew uninstall --force httpie
    $ brew install --HEAD httpie


Otherwise with ``pip``:

.. code-block:: bash

    $ python -m pip install --upgrade https://github.com/httpie/httpie/archive/master.tar.gz


Verify that now we have the
`current development version identifier <https://github.com/httpie/httpie/blob/master/httpie/__init__.py#L6>`_
with the ``-dev`` suffix, for example:

.. code-block:: bash

    $ http --version
    # 2.0.0-dev


Usage
=====


Hello World:


.. code-block:: bash

    $ https httpie.io/hello


Synopsis:

.. code-block:: bash

    $ http [flags] [METHOD] URL [ITEM [ITEM]]


See also ``http --help``.


Examples
--------

Custom `HTTP method`_, `HTTP headers`_ and `JSON`_ data:

.. code-block:: bash

    $ http PUT pie.dev/put X-API-Token:123 name=John


Submitting `forms`_:

.. code-block:: bash

    $ http -f POST pie.dev/post hello=World


See the request that is being sent using one of the `output options`_:

.. code-block:: bash

    $ http -v pie.dev/get


Build and print a request without sending it using `offline mode`_:

.. code-block:: bash

    $ http --offline pie.dev/post hello=offline


Use `GitHub API`_ to post a comment on an
`issue <https://github.com/httpie/httpie/issues/83>`_
with `authentication`_:

.. code-block:: bash

    $ http -a USERNAME POST https://api.github.com/repos/httpie/httpie/issues/83/comments body='HTTPie is awesome! :heart:'


Upload a file using `redirected input`_:

.. code-block:: bash

    $ http pie.dev/post < files/data.json


Download a file and save it via `redirected output`_:

.. code-block:: bash

    $ http pie.dev/image/png > image.png


Download a file ``wget`` style:

.. code-block:: bash

    $ http --download pie.dev/image/png

Use named `sessions`_ to make certain aspects of the communication persistent
between requests to the same host:


.. code-block:: bash

    $ http --session=logged-in -a username:password pie.dev/get API-Key:123


.. code-block:: bash

    $ http --session=logged-in pie.dev/headers


Set a custom ``Host`` header to work around missing DNS records:

.. code-block:: bash

    $ http localhost:8000 Host:example.com

..


HTTP method
===========

The name of the HTTP method comes right before the URL argument:

.. code-block:: bash

    $ http DELETE pie.dev/delete


Which looks similar to the actual ``Request-Line`` that is sent:

.. code-block:: http

    DELETE /delete HTTP/1.1


When the ``METHOD`` argument is omitted from the command, HTTPie defaults to
either ``GET`` (with no request data) or ``POST`` (with request data).


Request URL
===========

The only information HTTPie needs to perform a request is a URL.

The default scheme is ``http://`` and can be omitted from the argument:

.. code-block:: bash

    $ http example.org
    # => http://example.org


HTTPie also installs an ``https`` executable, where the default
scheme is ``https://``:


.. code-block:: bash

    $ https example.org
    # => https://example.org


Querystring parameters
----------------------

If you find yourself manually constructing URLs with querystring parameters
on the terminal, you may appreciate the ``param==value`` syntax for appending
URL parameters.

With that, you don’t have to worry about escaping the ``&``
separators for your shell. Additionally, any special characters in the
parameter name or value get automatically URL-escaped
(as opposed to parameters specified in the full URL, which HTTPie doesn’t
modify).

.. code-block:: bash

    $ http https://api.github.com/search/repositories q==httpie per_page==1


.. code-block:: http

    GET /search/repositories?q=httpie&per_page=1 HTTP/1.1



URL shortcuts for ``localhost``
-------------------------------

Additionally, curl-like shorthand for localhost is supported.
This means that, for example ``:3000`` would expand to ``http://localhost:3000``
If the port is omitted, then port 80 is assumed.

.. code-block:: bash

    $ http :/foo


.. code-block:: http

    GET /foo HTTP/1.1
    Host: localhost


.. code-block:: bash

    $ http :3000/bar


.. code-block:: http

    GET /bar HTTP/1.1
    Host: localhost:3000


.. code-block:: bash

    $ http :


.. code-block:: http

    GET / HTTP/1.1
    Host: localhost


Other default schemes
---------------------

When HTTPie is invoked as ``https`` then the default scheme is ``https://``
(``$ https example.org`` will make a request to ``https://example.org``).

You can also use the ``--default-scheme <URL_SCHEME>`` option to create
shortcuts for other protocols than HTTP (possibly supported via plugins).
Example for the `httpie-unixsocket <https://github.com/httpie/httpie-unixsocket>`_ plugin:

.. code-block:: bash

    # Before
    $ http http+unix://%2Fvar%2Frun%2Fdocker.sock/info


.. code-block:: bash

    # Create an alias
    $ alias http-unix='http --default-scheme="http+unix"'


.. code-block:: bash

    # Now the scheme can be omitted
    $ http-unix %2Fvar%2Frun%2Fdocker.sock/info


``--path-as-is``
----------------

The standard behaviour of HTTP clients is to normalize the path portion of URLs by squashing dot segments
as a typically filesystem would:


.. code-block:: bash

    $ http -v example.org/./../../etc/password

.. code-block:: http

    GET /etc/password HTTP/1.1


The ``--path-as-is`` option allows you to disable this behavior:

.. code-block:: bash

    $ http --path-as-is -v example.org/./../../etc/password

.. code-block:: http

    GET /../../etc/password HTTP/1.1


Request items
=============

There are a few different *request item* types that provide a
convenient mechanism for specifying HTTP headers, simple JSON and
form data, files, and URL parameters.

They are key/value pairs specified after the URL. All have in
common that they become part of the actual request that is sent and that
their type is distinguished only by the separator used:

CHANGELOG.rst
==========
Change Log
==========

This document records all notable changes to `HTTPie <https://httpie.org>`_.
This project adheres to `Semantic Versioning <https://semver.org/>`_.



`2.5.0-dev`_ (unreleased)
-------------------------
* Fixed ``--continue --download`` with a single byte to be downloaded left. (`#1032`_)
* Added ``--request, -r`` to allow specifying the raw request body without extra processing whatsoever. (`#534`_)


`2.4.0`_ (2021-02-06)
---------------------
* Added support for ``--session`` cookie expiration based on ``Set-Cookie: max-age=<n>``. (`#1029`_)
* Show a ``--check-status`` warning with ``--quiet`` as well, not only when the output is redirected. (`#1026`_)
* Fixed upload with ``--session`` (`#1020`_).
* Fixed a missing blank line between request and response (`#1006`_).


`2.3.0`_ (2020-10-25)
-------------------------

* Added support for streamed uploads (`#201`_).
* Added support for multipart upload streaming (`#684`_).
* Added support for body-from-file upload streaming (``http pie.dev/post @file``).
* Added ``--chunked`` to enable chunked transfer encoding (`#753`_).
* Added ``--multipart`` to allow ``multipart/form-data`` encoding for non-file ``--form`` requests as well.
* Added support for preserving field order in multipart requests (`#903`_).
* Added ``--boundary`` to allow a custom boundary string for ``multipart/form-data`` requests.
* Added support for combining cookies specified on the CLI and in a session file (`#932`_).
* Added out of the box SOCKS support with no extra installation (`#904`_).
* Added ``--quiet, -q`` flag to enforce silent behaviour.
* Fixed the handling of invalid ``expires`` dates in ``Set-Cookie`` headers (`#963`_).
* Removed Tox testing entirely (`#943`_).


`2.2.0`_ (2020-06-18)
-------------------------

* Added support for custom content types for uploaded files (`#668`_).
* Added support for ``$XDG_CONFIG_HOME`` (`#920`_).
* Added support for ``Set-Cookie``-triggered cookie expiration (`#853`_).
* Added ``--format-options`` to allow disabling sorting, etc. (`#128`_)
* Added ``--sorted`` and ``--unsorted`` shortcuts for (un)setting all sorting-related ``--format-options``. (`#128`_)
* Added ``--ciphers`` to allow configuring OpenSSL ciphers (`#870`_).
* Added ``netrc`` support for auth plugins. Enabled for ``--auth-type=basic``
  and ``digest``, 3rd parties may opt in (`#718`_, `#719`_, `#852`_, `#934`_).
* Fixed built-in plugins-related circular imports (`#925`_).


`2.1.0`_ (2020-04-18)
---------------------

* Added ``--path-as-is`` to bypass dot segment (``/../`` or ``/./``)
  URL squashing (`#895`_).
* Changed the default ``Accept`` header value for JSON requests from
  ``application/json, */*`` to ``application/json, */*;q=0.5``
  to clearly indicate preference (`#488`_).
* Fixed ``--form`` file upload mixed with redirected ``stdin`` error handling
  (`#840`_).


`2.0.0`_ (2020-01-12)
-------------------------
* Removed Python 2.7 support (`EOL Jan 2020 <https://www.python.org/doc/sunset-python-2/>`_).
* Added ``--offline`` to allow building an HTTP request and printing it but not
  actually sending it over the network.
* Replaced the old collect-all-then-process handling of HTTP communication
  with one-by-one processing of each HTTP request or response as they become
  available. This means that you can see headers immediately,
  see what is being sent even if the request fails, etc.
* Removed automatic config file creation to avoid concurrency issues.
* Removed the default 30-second connection ``--timeout`` limit.
* Removed Python’s default limit of 100 response headers.
* Added ``--max-headers`` to allow setting the max header limit.
* Added ``--compress`` to allow request body compression.
* Added ``--ignore-netrc`` to allow bypassing credentials from ``.netrc``.
* Added ``https`` alias command with ``https://`` as the default scheme.
* Added ``$ALL_PROXY`` documentation.
* Added type annotations throughout the codebase.
* Added ``tests/`` to the PyPi package for the convenience of
  downstream package maintainers.
* Fixed an error when ``stdin`` was a closed fd.
* Improved ``--debug`` output formatting.


`1.0.3`_ (2019-08-26)
---------------------

* Fixed CVE-2019-10751 — the way the output filename is generated for
  ``--download`` requests without ``--output`` resulting in a redirect has
  been changed to only consider the initial URL as the base for the generated
  filename, and not the final one. This fixes a potential security issue under
  the following scenario:

  1. A ``--download`` request with no explicit ``--output`` is made (e.g.,
     ``$ http -d example.org/file.txt``), instructing httpie to

httpie/cli/definition.py
"""
CLI arguments definition.

"""
from argparse import (FileType, OPTIONAL, SUPPRESS, ZERO_OR_MORE)
from textwrap import dedent, wrap

from .. import __doc__, __version__
from .argparser import HTTPieArgumentParser
from .argtypes import (
    KeyValueArgType, SessionNameValidator,
    readable_file_arg,
)
from .constants import (
    DEFAULT_FORMAT_OPTIONS, OUTPUT_OPTIONS,
    OUTPUT_OPTIONS_DEFAULT, OUT_REQ_BODY, OUT_REQ_HEAD,
    OUT_RESP_BODY, OUT_RESP_HEAD, PRETTY_MAP, PRETTY_STDOUT_TTY_ONLY,
    RequestType, SEPARATOR_GROUP_ALL_ITEMS, SEPARATOR_PROXY,
    SORTED_FORMAT_OPTIONS_STRING,
    UNSORTED_FORMAT_OPTIONS_STRING,
)
from ..output.formatters.colors import (
    AUTO_STYLE, AVAILABLE_STYLES, DEFAULT_STYLE,
)
from ..plugins.builtin import BuiltinAuthPlugin
from ..plugins.registry import plugin_manager
from ..sessions import DEFAULT_SESSIONS_DIR
from ..ssl import AVAILABLE_SSL_VERSION_ARG_MAPPING, DEFAULT_SSL_CIPHERS


parser = HTTPieArgumentParser(
    prog='http',
    description='%s <https://httpie.org>' % __doc__.strip(),
    epilog=dedent('''
    For every --OPTION there is also a --no-OPTION that reverts OPTION
    to its default value.

    Suggestions and bug reports are greatly appreciated:

        https://github.com/httpie/httpie/issues

    '''),
)

#######################################################################
# Positional arguments.
#######################################################################

positional = parser.add_argument_group(
    title='Positional Arguments',
    description=dedent('''
    These arguments come after any flags and in the order they are listed here.
    Only URL is required.

    ''')
)
positional.add_argument(
    dest='method',
    metavar='METHOD',
    nargs=OPTIONAL,
    default=None,
    help='''
    The HTTP method to be used for the request (GET, POST, PUT, DELETE, ...).

    This argument can be omitted in which case HTTPie will use POST if there
    is some data to be sent, otherwise GET:

        $ http example.org               # => GET
        $ http example.org hello=world   # => POST

    '''
)
positional.add_argument(
    dest='url',
    metavar='URL',
    help='''
    The scheme defaults to 'http://' if the URL does not include one.
    (You can override this with: --default-scheme=https)

    You can also use a shorthand for localhost

        $ http :3000                    # => http://localhost:3000
        $ http :/foo                    # => http://localhost/foo

    '''
)
positional.add_argument(
    dest='request_items',
    metavar='REQUEST_ITEM',
    nargs=ZERO_OR_MORE,
    default=None,
    type=KeyValueArgType(*SEPARATOR_GROUP_ALL_ITEMS),
    help=r'''
    Optional key-value pairs to be included in the request. The separator used
    determines the type:

    ':' HTTP headers:

        Referer:http://httpie.org  Cookie:foo=bar  User-Agent:bacon/1.0

    '==' URL parameters to be appended to the request URI:

        search==httpie

    '=' Data fields to be serialized into a JSON object (with --json, -j)
        or form data (with --form, -f):

        name=HTTPie  language=Python  description='CLI HTTP client'

    ':=' Non-string JSON data fields (only with --json, -j):

        awesome:=true  amount:=42  colors:='["red", "green", "blue"]'

    '@' Form file fields (only with --form or --multipart):

        cv@~/Documents/CV.pdf
        cv@'~/Documents/CV.pdf;type=application/pdf'

    '=@' A data field like '=', but takes a file path and embeds its content:

         essay=@Documents/essay.txt

    ':=@' A raw JSON field like ':=', but takes a file path and embeds its content:

        package:=@./package.json

    You can use a backslash to escape a colliding separator in the field name:

        field-name-with\:colon=value

    '''
)

#######################################################################
# Content type.
#######################################################################

content_type = parser.add_argument_group(
    title='Predefined Content Types',
    description=None
)

content_type.add_argument(
    '--json', '-j',
    action='store_const',
    const=RequestType.JSON,
    dest='request_type',
    help='''
    (default) Data items from the command line are serialized as a JSON object.
    The Content-Type and Accept headers are set to application/json
    (if not specified).

    '''
)
content_type.add_argument(
    '--form', '-f',
    action='store_const',
    const=RequestType.FORM,
    dest='request_type',
    help='''
    Data items from the command line are serialized as form fields.

    The Content-Type is set to application/x-www-form-urlencoded (if not
    specified). The presence of any file fields results in a
    multipart/form-data request.

    '''
)
content_type.add_argument(
    '--multipart',
    action='store_const',
    const=RequestType.MULTIPART,
    dest='request_type',
    help='''
    Similar to --form, but always sends a multipart/form-data
    request (i.e., even without files).

    '''
)
content_type.add_argument(
    '--boundary',
    help='''
    Specify a custom boundary string for multipart/form-data requests.
    Only has effect only together with --form.

    '''
)
content_type.add_argument(
    '--request', '-r',
    help='''
    Specify the raw request body without extra processing whatsoever.

    '''
)


#######################################################################
# Content processing.
#######################################################################

content_processing = parser.add_argument_group(
    title='Content Processing Options',
    description=None
)

content_processing.add_argument(
    '--compress', '-x',
    action='count',
    default=0,
    help='''
    Content compressed (encoded) with Deflate algorithm.
    The Content-Encoding header is set to deflate.

    Compression is skipped if it appears that compression ratio is
    negative. Compression can be forced by repeating the argument.

    '''
)

#######################################################################
# Output processing
#######################################################################

output_processing = parser.add_argument_group(title='Output Processing')

output_processing.add_argument(
    '--pretty',
    dest='prettify',
    default=PRETTY_STDOUT_TTY_ONLY,
    choices=sorted(PRETTY_MAP.keys()),
    help='''
    Controls output processing. The value can be "none" to not prettify
    the output (default for redirected output), "all" to apply both colors
    and formatting (default for terminal output), "colors", or "format".

    '''
)
output_processing.add_argument(
    '--style', '-s',
    dest='style',
    metavar='STYLE',
    default=DEFAULT_STYLE,
    choices=AVAILABLE_STYLES,
    help='''
    Output coloring style (default is "{default}"). It can be One of:

    {available_styles}

    The "{auto_style}" style follows your terminal's ANSI color styles.

    For non-{auto_style} styles to work properly, please make sure that the
    $TERM environment variable is set to "xterm-256color" or similar
    (e.g., via `export TERM=xterm-256color' in your ~/.bashrc).

    '''.format(
        default=DEFAULT_STYLE,
        available_styles='\n'.join(
            '{0}{1}'.format(8 * ' ', line.strip())
            for line in wrap(', '.join(sorted(AVAILABLE_STYLES)), 60)
        ).strip(),
        auto_style=AUTO_STYLE,
    )
)
_sorted_kwargs = {
    'action': 'append_const',
    'const': SORTED_FORMAT_OPTIONS_STRING,
    'dest': 'format_options'
}
_unsorted_kwargs = {
    'action': 'append_const',
    'const': UNSORTED_FORMAT_OPTIONS_STRING,
    'dest': 'format_options'
}
# The closest approx. of the documented resetting to default via --no-<option>.
# We hide them from the doc because they act only as low-level aliases here.
output_processing.add_argument('--no-unsorted', **_sorted_kwargs, help=SUPPRESS)
output_processing.add_argument('--no-sorted', **_unsorted_kwargs, help=SUPPRESS)

output_processing.add_argument(
    '--unsorted',
    **_unsorted_kwargs,
    help=f'''
    Disables all sorting while formatting output. It is a shortcut for:

        --format-options={UNSORTED_FORMAT_OPTIONS_STRING}

    '''
)
output_processing.add_argument(
    '--sorted',
    **_sorted_kwargs,
    help=f'''
    Re-enables all sorting options while formatting output. It is a shortcut for:

        --format-options={SORTED_FORMAT_OPTIONS_STRING}

    '''
)


output_processing.add_argument(
    '--format-options',
    action='append',
    help='''
    Controls output formatting. Only relevant when formatting is enabled
    through (explicit or implied) --pretty=all or --pretty=format.
    The following are the default options:

        {option_list}

    You may use this option multiple times, as well as specify multiple
    comma-separated options at the same time. For example, this modifies the
    settings to disable the sorting of JSON keys, and sets the indent size to 2:

        --format-options json.sort_keys:false,json.indent:2

    This is something you will typically put into your config file.

    '''.format(
        option_list='\n'.join(
            (8 * ' ') + option for option in DEFAULT_FORMAT_OPTIONS).strip()
    )
)

#######################################################################
# Output options
#######################################################################
output_options = parser.add_argument_group(title='Output Options')

output_options.add_argument(
    '--print', '-p',
    dest='output_options',
    metavar='WHAT',
    help=f'''
    String specifying what the output should contain:

        '{OUT_REQ_HEAD}' request headers
        '{OUT_REQ_BODY}' request body
        '{OUT_RESP_HEAD}' response headers
        '{OUT_RESP_BODY}' response body

    The default behaviour is '{OUTPUT_OPTIONS_DEFAULT}' (i.e., the response
    headers and body is printed), if standard output is not redirected.
    If the output is piped to another program or to a file, then only the
    response body is printed by default.

    '''
)
output_options.add_argument(
    '--headers', '-h',
    dest='output_options',
    action='store_const',
    const=OUT_RESP_HEAD,
    help=f'''
    Print only the response headers. Shortcut for --print={OUT_RESP_HEAD}.

    '''
)
output_options.add_argument(
    '--body', '-b',
    dest='output_options',
    action='store_const',
    const=OUT_RESP_BODY,
    help=f'''
    Print only the response body. Shortcut for --print={OUT_RESP_BODY}.

    '''
)

output_options.add_argument(
    '--verbose', '-v',
    dest='verbose',
    action='store_true',
    help='''
    Verbose output. Print the whole request as well as the response. Also print
    any intermediary requests/responses (such as redirects).
    It's a shortcut for: --all --print={0}

    '''.format(''.join(OUTPUT_OPTIONS))
)
output_options.add_argument(
    '--all',
    default=False,
    action='store_true',
    help='''
    By default, only the final request/response is shown. Use this flag to show
    any intermediary requests/responses as well. Intermediary requests include
    followed redirects (with --follow), the first unauthorized request when
    Digest auth is used (--auth=digest), etc.

    '''
)
output_options.add_argument(
    '--history-print', '-P',
    dest='output_options_history',
    metavar='WHAT',
    help='''
    The same as --print, -p but applies only to intermediary requests/responses
    (such as redirects) when their inclusion is enabled with --all. If this
    options is not specified, then they are formatted the same way as the final
    response.

    '''
)
output_options.add_argument(
    '--stream', '-S',
    action='store_true',
    default=False,
    help='''
    Always stream the response body by line, i.e., behave like `tail -f'.

    Without --stream and with --pretty (either set or implied),
    HTTPie fetches the whole response before it outputs the processed data.

    Set this option when you want to continuously display a prettified
    long-lived response, such as one from the Twitter streaming API.

    It is useful also without --pretty: It ensures that the output is flushed
    more often and in smaller chunks.

    '''
)
output_options.add_argument(
    '--output', '-o',
    type=FileType('a+b'),
    dest='output_file',
    metavar='FILE',
    help='''
    Save output to FILE instead of stdout. If --download is also set, then only
    the response body is saved to FILE. Other parts of the HTTP exchange are
    printed to stderr.

    '''

)

output_options.add_argument(
    '--download', '-d',
    action='store_true',
    default=False,
    help='''
    Do not print the response body to stdout. Rather, download it and store it
    in a file. The filename is guessed unless specified with --output
    [filename]. This action is similar to the default behaviour of wget.

    '''
)

output_options.add_argument(
    '--continue', '-c',
    dest='download_resume',
    action='store_true',
    default=False,
    help='''
    Resume an interrupted download. Note that the --output option needs to be
    specified as well.

    '''
)

output_options.add_argument(
    '--quiet', '-q',
    action='store_true',
    default=False,
    help='''
    Do not print to stdout or stderr.
    stdout is still redirected if --output is specified.
    Flag doesn't affect behaviour of download beyond not printing to terminal.
    '''
)

#######################################################################
# Sessions
#######################################################################

sessions = parser.add_argument_group(title='Sessions') \
    .add_mutually_exclusive_group(required=False)

session_name_validator = SessionNameValidator(
    'Session name contains invalid characters.'
)

sessions.add_argument(
    '--session',
    metavar='SESSION_NAME_OR_PATH',
    type=session_name_validator,
    help=f'''
    Create, or reuse and update a session. Within a session, custom headers,
    auth credential, as well as any cookies sent by the server persist between
    requests.

    Session files are stored in:

        {DEFAULT_SESSIONS_DIR}/<HOST>/<SESSION_NAME>.json.

    '''
)
sessions.add_argument(
    '--session-read-only',
    metavar='SESSION_NAME_OR_PATH',
    type=session_name_validator,
    help='''
    Create or read a session without updating it form the request/response
    exchange.

    '''
)

#######################################################################
# Authentication
#######################################################################

# ``requests.request`` keyword arguments.
auth = parser.add_argument_group(title='Authentication')
auth.add_argument(
    '--auth', '-a',
    default=None,
    metavar='USER[:PASS]',
    help='''
    If only the username is provided (-a username), HTTPie will prompt
    for the password.

    ''',
)


class _AuthTypeLazyChoices:
    # Needed for plugin testing

    def __contains__(self, item):
        return item in plugin_manager.get_auth_plugin_mapping()

    def __iter__(self):
        return iter(sorted(plugin_manager.get_auth_plugin_mapping().keys()))


_auth_plugins = plugin_manager.get_auth_plugins()
auth.add_argument(
    '--auth-type', '-A',
    choices=_AuthTypeLazyChoices(),
    default=None,
    help='''
    The authentication mechanism to be used. Defaults to "{default}".

    {types}

    '''.format(default=_auth_plugins[0].auth_type, types='\n    '.join(
        '"{type}": {name}{package}{description}'.format(
            type=plugin.auth_type,
            name=plugin.name,
            package=(
                '' if issubclass(plugin, BuiltinAuthPlugin)
                else ' (provided by %s)' % plugin.package_name
            ),
            description=(
                '' if not plugin.description else
                '\n      ' + ('\n      '.join(wrap(plugin.description)))
            )
        )
        for plugin in _auth_plugins
    )),
)
auth.add_argument(
    '--ignore-netrc',
    default=False,
    action='store_true',
    help='''
    Ignore credentials from .netrc.

    ''',
)

#######################################################################
# Network
#######################################################################

network = parser.add_argument_group(title='Network')

network.add_argument(
    '--offline',
    default=False,
    action='store_true',
    help='''
    Build the request and print it but don’t actually send it.
    '''
)
network.add_argument(
    '--proxy',
    default=[],
    action='append',
    metavar='PROTOCOL:PROXY_URL',
    type=KeyValueArgType(SEPARATOR_PROXY),
    help='''
    String mapping protocol to the URL of the proxy
    (e.g. http:http://foo.bar:3128). You can specify multiple proxies with
    different protocols. The environment variables $ALL_PROXY, $HTTP_PROXY,
    and $HTTPS_proxy are supported as well.

    '''
)
network.add_argument(
    '--follow', '-F',
    default=False,
    action='store_true',
    help='''
    Follow 30x Location redirects.

    '''
)

network.add_argument(
    '--max-redirects',
    type=int,
    default=30,
    help='''
    By default, requests have a limit of 30 redirects (works with --follow).

    '''
)

network.add_argument(
    '--max-headers',
    type=int,
    default=0,
    help='''
    The maximum number of response headers to be read before giving up
    (default 0, i.e., no limit).

    '''
)

network.add_argument(
    '--timeout',
    type=float,
    default=0,
    metavar='SECONDS',
    help='''
    The connection timeout of the request in seconds.
    The default value is 0, i.e., there is no timeout limit.
    This is not a time limit on the entire response download;
    rather, an error is reported if the server has not issued a response for
    timeout seconds (more precisely, if no bytes have been received on
    the underlying socket for timeout seconds).

    '''
)
network.add_argument(
    '--check-status',
    default=False,
    action='store_true',
    help='''
    By default, HTTPie exits with 0 when no network or other fatal errors
    occur. This flag instructs HTTPie to also check the HTTP status code and
    exit with an error if the status indicates one.

    When the server replies with a 4xx (Client Error) or 5xx (Server Error)
    status code, HTTPie exits with 4 or 5 respectively. If the response is a
    3xx (Redirect) and --follow hasn't been set, then the exit status is 3.
    Also an error message is written to stderr if stdout is redirected.

    '''
)
network.add_argument(
    '--path-as-is',
    default=False,
    action='store_true',
    help='''
    Bypass dot segment (/../ or /./) URL squashing.

    '''
)

network.add_argument(
    '--chunked',
    default=False,
    action='store_true',
    help="""

    """
)

#######################################################################
# SSL
#######################################################################

ssl = parser.add_argument_group(title='SSL')
ssl.add_argument(
    '--verify',
    default='yes',
    help='''
    Set to "no" (or "false") to skip checking the host's SSL certificate.
    Defaults to "yes" ("true"). You can also pass the path to a CA_BUNDLE file
    for private certs. (Or you can set the REQUESTS_CA_BUNDLE environment
    variable instead.)
    '''
)
ssl.add_argument(
    '--ssl',
    dest='ssl_version',
    choices=list(sorted(AVAILABLE_SSL_VERSION_ARG_MAPPING.keys())),
    help='''
    The desired protocol version to use. This will default to
    SSL v2.3 which will negotiate the highest protocol that both
    the server and your installation of OpenSSL support. Available protocols
    may vary depending on OpenSSL installation (only the supported ones
    are shown here).

    '''
)
ssl.add_argument(
    '--ciphers',
    help=f'''

    A string in the OpenSSL cipher list format. By default, the following
    is used:

    {DEFAULT_SSL_CIPHERS}

    '''
)
ssl.add_argument(
    '--cert',
    default=None,
    type=readable_file_arg,
    help='''
    You can specify a local cert to use as client side SSL certificate.
    This file may either contain both private key and certificate or you may
    specify --cert-key separately.

    '''
)

ssl.add_argument(
    '--cert-key',
    default=None,
    type=readable_file_arg,
    help='''
    The private key to use with SSL. Only needed if --cert is given and the
    certificate file does not contain the private key.

    '''
)

#######################################################################
# Troubleshooting
#######################################################################

troubleshooting = parser.add_argument_group(title='Troubleshooting')

troubleshooting.add_argument(
    '--ignore-stdin', '-I',
    action='store_true',
    default=False,
    help='''
    Do not attempt to read stdin.

    '''
)
troubleshooting.add_argument(
    '--help',
    action='help',
    default=SUPPRESS,
    help='''
    Show this help message and exit.

    '''
)
troubleshooting.add_argument(
    '--version',
    action='version',
    version=__version__,
    help='''
    Show version and exit.

    '''
)
troubleshooting.add_argument(
    '--traceback',
    action='store_true',
    default=False,
    help='''
    Prints the exception traceback should one occur.

    '''
)
troubleshooting.add_argument(
    '--default-scheme',
    default="http",
    help='''
    The default scheme to use if not specified in the URL.

    '''
)
troubleshooting.add_argument(
    '--debug',
    action='store_true',
    default=False,
    help='''
    Prints the exception traceback should one occur, as well as other
    information useful for debugging HTTPie itself and for reporting bugs.

    '''
)
